### PR TITLE
feat(coil-extension): fire monetizationprogress event on tipping

### LINKED ISSUE #900: Tipping should trigger 'monetizationprogress' event?
Currently, leaving a tip issues a non-standard event `tip`. (Which as a side-note should probably have a vendor prefix since it's non-standard, like `coilTip` or `coiltip`. For an example how different browser vendors handle vendor prefixes for JS events, see [CSS3 Animation Events](https://www.sitepoint.com/css3-animation-javascript-event-handlers/).)

I think it's fine to have a separate event for tipping but I think a monetizationprogress event should be triggered also. For instance, if a page is unlocking a reward when I pay a certain amount of money, my tips should count towards that.

That does require that the receipts for the tip are somehow passed from the Coil tipping service back to the client I suppose?

Lmk if you think I'm way off here or - if you agree - how we might implement this.

### PR DESCRIPTION
**Pull request by sabinebertram**
Closes #900, extension side of https://github.com/coilhq/coil/pull/4490

- [x] Be sure to send a progress event per receipt returned via franklin
- [x] Fix @coil/client modifications
- [x] Deal with modified polyfill ( "tip" event is removed )

### CODE REVIEW

File: packages/coil-extension/src/background/services/BackgroundScript.ts [hunk 1]
```diff
@@ -728,6 +728,7 @@ export class BackgroundScript {
         command: 'tip',
         data: {
           paymentPointer: receiver,
+          requestId: streamId,
```
Comment (sublimator) [inline packages/coil-extension/src/background/services/BackgroundScript.ts:731]: This implies the SPSP server not really doing anything relying on the Web-Monetization-Id in terms of accounting ?
It's already been deprecated right

File: tsconfig.json [hunk 2]
```diff
@@ -1,50 +1,27 @@
...
-      "@coil/privacypass-sjcl": [
-        "packages/coil-privacypass-sjcl/src"
-      ],
-      "@coil/webpack-utils": [
-        "packages/coil-webpack-utils/src"
-      ],
-      "@dier-makr/annotations": [
-        "packages/dier-makr-annotations/src"
-      ],
+      "@coil/monorepo-upkeep": ["packages/coil-monorepo-upkeep/src"],
```
Comment (sublimator) [inline tsconfig.json:8]: run `yarn upkeep`, it should format these in the canonical format expected by CI

File: packages/coil-extension/src/background/services/BackgroundScript.ts (hunk 1 above)
Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: Should the requestId be sent to the Coil backend to send as Web-Monetization-Id to the SPSP server when sending the tip? For backend integrations that aren't yet using receipts ?

File: packages/coil-extension/src/content/services/ContentScript.ts [hunk 3]
```diff
@@ -161,13 +161,14 @@ export class ContentScript {
           this.onFrameAllowedChanged(request)
         } else if (request.command === 'tip') {
           debug('sendTip event')
-          const detail: TipEvent['detail'] = {
+          const detail: MonetizationProgressEvent['detail'] = {
             amount: request.data.amount,
             assetCode: request.data.assetCode,
             assetScale: request.data.assetScale,
-            paymentPointer: request.data.paymentPointer
+            paymentPointer: request.data.paymentPointer,
```
Comment (sublimator) [inline packages/coil-extension/src/content/services/ContentScript.ts:168]: Backend not yet sending receipts ?

File: tsconfig.json (hunk 2 above)
Comment (sublimator) [reply tsconfig.json]: pushed !

Comment (sabinebertram) [reply tsconfig.json]: Thanks! I was wondering what I need to do to fix this issue.

File: packages/coil-extension/src/content/services/ContentScript.ts (hunk 3 above)
Comment (sabinebertram) [reply packages/coil-extension/src/content/services/ContentScript.ts:168]: We are not passing them through yet. I'm tackling that next.

File: packages/coil-extension/src/background/services/BackgroundScript.ts (hunk 1 above)
Comment (sabinebertram) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: I'm afraid I don't understand what you are asking.

Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: Sorry, bad sleep! I'm going to go with that :)

Depending on the WM integration, it may be that the backend tracks balances via the requestId sent to the SPSP server as "Web-Monetization-Id", so this means that in some cases, checking the balance on the backend by requestId won't show the tip. Sigh, I realize that's a bit vague, but that's because we're talking about potential integrations we can't really point to and discuss?

Comment (sharafian) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: It sounds like the potential issue here in most cases we have `Web-Monetization-Id` sent on the SPSP request when we make payments, and then it's emitted on the monetizationprogress event so if the server wants to correlate SPSP request/payment to the monetizationprogress event, it can.

However, Franklin doesn't send the `Web-Monetization-Id` header, so emitting a requestId on this payment may be a bit disingenuous. The site may try to cross reference that requestId with a payment only to find that no SPSP requests were made with the `Web-Monetization-Id` header.

But with the `Web-Monetization-Id` feature being deprecated (in favor of submitting receipts to verify payment) we could possibly include the requestId to avoid breaking the interface without actually sending it, with the assumption nobody will check that `Web-Monetization-Id` was really sent.

It might be better, though, just to be thorough and actually allow the `Web-Monetization-Id`/requestId to be passed to the API, then to Franklin, and then include it in Franklin's SPSP request. I don't think it would be very hard to add as an optional field and it's one less thing to worry about.

Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: Ok, I will hold off on merging then

**Review by sublimator — CHANGES_REQUESTED** [review]
Holding until backend graphql mutation returns receipts and takes an optional requestId

File: packages/coil-extension/src/background/services/BackgroundScript.ts (hunk 1 above)
Comment (sabinebertram) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: > we could possibly include the requestId to avoid breaking the interface without actually sending it

Include it in the `monetizationprogress` event without sending it via Franklin in the SPSP request? Just making sure I get it.

> It might be better, though, just to be thorough and actually allow the Web-Monetization-Id/requestId to be passed to the API, then to Franklin, and then include it in Franklin's SPSP request.

It sounds cleaner to me.

Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: > Include it in the monetizationprogress event without sending it via Franklin

That's basically the state of the PR now (if I'm not confusing myself)

Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:731]: > Ok, I will hold off on merging then

@sharafian 
With an eye ((perhaps foolishly optimistic)) to getting a release out before the year's end?
With the fix for #1359 (and in that vein, some fireworks 31/12) and maybe #1167 now that @fruehle is back from holidays

**Comment by sabinebertram** [comment]
Depends on coil api changes to be merged ([#4490](https://github.com/coilhq/coil/pull/4490)).

File: packages/coil-extension/src/content/services/ContentScript.ts [hunk 4]
```diff
@@ -161,13 +161,15 @@ export class ContentScript {
...
+            requestId: request.data.requestId,
+            receipt: request.data.receipts ? request.data.receipts[0] : ''
```
[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/src/content/services/ContentScript.ts:170]: @wilsonianb 
For some reason I'm remembering you as the person who altered Franklin to return receipts?

**Review by sublimator — CHANGES_REQUESTED** [review]

File: packages/coil-extension/src/content/services/ContentScript.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wilsonianb) [reply packages/coil-extension/src/content/services/ContentScript.ts:170]: Yeah, I think the intention was for there to be a `monetizationprogress` emitted for each returned receipt
https://github.com/coilhq/coil/pull/4304#issuecomment-722018581

[the remark this answers is not in the document]
Comment (sublimator) [reply packages/coil-extension/src/content/services/ContentScript.ts:170]: @wilsonianb 

Great, thanks for the confirm/info

File: packages/web-monetization-wext/src/content/wmPolyfill.ts [hunk 5]
```diff
@@ -1,7 +1,8 @@
 const events = [
   'monetizationstart',
   'monetizationstop',
-  'monetizationpending'
+  'monetizationpending',
```
Comment (sublimator) [inline packages/web-monetization-wext/src/content/wmPolyfill.ts]: need a CI test that checks the manifest.json hash listed in the CSP matches the hashed polyfill

File: packages/coil-extension/src/background/services/BackgroundScript.ts [hunk 6]
```diff
@@ -724,36 +734,50 @@ export class BackgroundScript {
     }
 
     const receiver = stream.getPaymentPointer()
-    const { assetCode, assetScale, exchangeRate } = stream.getAssetDetails()
```
Comment (sublimator) [inline packages/coil-extension/src/background/services/BackgroundScript.ts:727]: See: https://github.com/coilhq/web-monetization-projects/pull/567#discussion_r419924429

Removed as the new tipping endpoint will return the exact amount sent

File: packages/coil-extension/src/content/services/ContentScript.ts [hunk 7]
```diff
@@ -160,14 +160,26 @@ export class ContentScript {
         } else if (request.command === 'onFrameAllowedChanged') {
           this.onFrameAllowedChanged(request)
         } else if (request.command === 'tip') {
-          debug('sendTip event')
-          const detail: TipEvent['detail'] = {
+          debug('sendTip event', request.data)
+          const detail: MonetizationProgressEvent['detail'] = {
```
Comment (sublimator) [inline packages/coil-extension/src/content/services/ContentScript.ts:164]: image

This doesn't seem right

Comment (sublimator) [reply packages/coil-extension/src/content/services/ContentScript.ts:164]: https://github.com/coilhq/coil/pull/4490/files#r589991430

File: packages/coil-extension/src/background/services/BackgroundScript.ts [hunk 8]
```diff
@@ -724,36 +734,63 @@ export class BackgroundScript {
...
+              }
             }
           }
         `,
         token,
         variables: {
-          receiver
+          receiver,
+          amount: undefined, // default
+          webMonetizationId: streamId
```
Comment (wilsonianb) [inline packages/coil-extension/src/background/services/BackgroundScript.ts:761]: Would `streamId` be the same `requestId` used for web monetization's `Web-Monetization-Id`?

File: packages/coil-client/src/graphQlClient.ts [hunk 9]
```diff
@@ -38,7 +38,7 @@ export class GraphQlClient {
 
   public constructor(
     @inject(GraphQlClientOptions)
-    private config: GraphQlClientOptions = new GraphQlClientOptions()
+    protected config: GraphQlClientOptions = new GraphQlClientOptions()
```
Comment (wilsonianb) [inline packages/coil-client/src/graphQlClient.ts:41]: Is this change needed?

**Review by wilsonianb — APPROVED** [review]
LGTM :+1:

File: packages/coil-client/src/graphQlClient.ts (hunk 9 above)
Comment (sublimator) [reply packages/coil-client/src/graphQlClient.ts:41]: no, not needed, but it's not "incorrect"
it was cause I'd modified these files previously in another branch
so I just pinched the changes
I'm not religious about confining small changes to their own commits

File: packages/coil-extension/src/background/services/BackgroundScript.ts (hunk 8 above)
Comment (sublimator) [reply packages/coil-extension/src/background/services/BackgroundScript.ts:761]: Yep, streamId is what will be used for WMID and the requestId in the events

**Comment by sublimator** [comment]
Closing, as Bernie won't support progress events due to async sending of funds:
https://github.com/coilhq/coil/pull/4714#issuecomment-828221546
